[lexical-website] Documentation Update: Add detailed guides for key and clone concepts

kirandash opened the pull request:
## Description
### Current behavior:
- Documentation about key management and node cloning is not sufficiently detailed
- Users often misunderstand when to use `__key` and `clone` methods (see #6937)
- Need clearer guidance on internal vs public APIs

### Changes in this PR:
1. Added `key-management.md`:
   - Comprehensive guide about Lexical's key system
   - Clear examples of correct/incorrect key usage
   - Best practices and common pitfalls
   - Performance implications
   - Testing considerations
2. Added `node-cloning.md`:
   - Detailed explanation of different cloning mechanisms
   - Clear distinction between internal (`static clone()`) and public APIs (`$copyNode()`, `getWritable()`)
   - Best practices for each cloning method
   - Performance considerations
   - Testing examples
3. Updated sidebar navigation to include both new guides

These new guides provide clear, focused documentation that helps users understand:
- When and how to use `__key` properly
- The difference between internal and public cloning APIs
- Best practices for both key management and node cloning

Closes #6937

## Test plan

### Before

- Insufficient documentation about key and clone usage
- Users confused about proper API usage
- Lack of clear examples and best practices


### After

- Two comprehensive, focused guides
- Clear examples of correct/incorrect usage
- Detailed best practices and performance considerations
- Better organized information with specific use cases

etrepum reviewed (commented):
This is a great start, a few details need to be corrected but I think this documentation will help a lot of people once it's wrapped up!

kirandash commented:
Thanks for the review @etrepum! I've addressed all the review feedbacks:

Please let me know if any further changes are needed!

etrepum commented on packages/lexical-website/docs/concepts/key-management.md:
```diff
@@ -0,0 +1,214 @@
+  __prev: null | NodeKey,
+  __next: null | NodeKey,
+  __parent: null | NodeKey,
+  __firstChild: null | NodeKey,
+  __lastChild: null | NodeKey
```
```suggestion
  __first: null | NodeKey,
  __last: null | NodeKey
```

etrepum commented on packages/lexical-website/docs/concepts/key-management.md:
```diff
@@ -0,0 +1,214 @@
+   }
+
+   // ✅ Correct implementation
+   class MyCustomNode extends ElementNode {
+     constructor(someData: string, key?: NodeKey) {
+       super(key);
+       this.__someData = someData;
+     }
+     
+     static clone(node: MyCustomNode): MyCustomNode {
+       return new MyCustomNode(node.__someData, node.__key);
+     }
+   }
```
For completeness we probably want to demonstrate full subclass

```suggestion
   class MyCustomNode extends ElementNode {
     __someData: string;
     constructor(someData: string, key?: NodeKey) {
       super(key);
       this.__someData = someData;
     }
     
     static clone(node: MyCustomNode): MyCustomNode {
       return new MyCustomNode(node.__someData, node.__key);
     }
     
     afterCloneFrom(prevNode: this): void {
       super.afterCloneFrom(this);
       this.__someData = prevNode.__someData;
     }
   }
```

etrepum commented on packages/lexical-website/docs/concepts/key-management.md:
````diff
@@ -0,0 +1,214 @@
+   }
+   ```
+
+4. **Key Reuse in replace Function**
+   ```typescript
+   // ❌ Never do this - reusing key for different node type
+   class MyCustomNode extends ElementNode {
+     replace(replaceWith: LexicalNode) {
+       return new DifferentNodeType(data, this.__key); // Wrong: reusing key for different node type
+     }
+   }
+
+   // ✅ Correct: Let Lexical handle the key
+   class MyCustomNode extends ElementNode {
+     replace(replaceWith: LexicalNode) {
+       return new DifferentNodeType(data); // Correct: Lexical will assign a new key
+     }
+   }
+   ```
````
This is an entirely different replace that should not be implemented in this way, see node overrides/node replacement https://lexical.dev/docs/concepts/node-replacement

Not sure why this doc page is missing items from the table of contents so I can't permalink to that section directly, must be a formatting mistake in the markdown.

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
```diff
@@ -0,0 +1,169 @@
+
+### What is `clone`?
+
+The `clone` method is a static method required by all Lexical nodes that creates a copy of a node. However, it's important to understand that this is an **internal API** used by Lexical's state management system.
+
```
```suggestion
The `clone` method is a static method required by all Lexical nodes that creates the next version of a node. However, it's important to understand that this is an **internal API** used by Lexical's state management system.

```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,169 @@
+   ```typescript
+   // Example of a node method implementation
+   class MyCustomNode extends ElementNode {
+     setSomeData(data: string): void {
+       this.getWritable().__data = data;
+     }
````
```suggestion
     setData(data: string): this {
       const self = this.getWritable();
       self.__data = data;
       return self;
     }
```

Conventionally setters are written in a chainable style

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,169 @@
+```typescript
+// ✅ Correct: Implementation of a node method
+class MyCustomNode extends ElementNode {
+  setData(data: string): void {
+    this.getWritable().__data = data;
+  }
````
```suggestion
  setData(data: string): this {
    const self = this.getWritable();
    self.__data = data;
    return self;
  }
```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,169 @@
+2. **State Management**
+   ```typescript
+   class MyCustomNode extends ElementNode {
+     setData(data: string): void {
+       this.getWritable().__data = data; // Properly tracked by editor
+     }
````
```suggestion
     setData(data: string): this {
       const self = this.getWritable();
       self.__data = data; // Properly tracked by editor
       return self;
     }
```

etrepum reviewed (commented):
Very close, the key reuse in replace function section needs to be rewritten. The rest of the changes are superficial

kirandash commented:
Thanks for the detailed review @etrepum! I've addressed all your feedback:

1. [Fixed internal node pointer names](https://github.com/facebook/lexical/pull/7448/commits/174bc738dd5831b460e6a662c4e20babf5c7c35f):
   - Changed `__firstChild` and `__lastChild` to `__first` and `__last`

2. [Added complete node subclass example](https://github.com/facebook/lexical/pull/7448/commits/ca585d0c588b72bfea3820f4d0df1a80c500bc3d)

3. [Updated node method examples to show chainable style](https://github.com/facebook/lexical/pull/7448/commits/fbd5f75670708e5837313c8a66825cfee6aacc34)

4. [Clarified `clone()` description](https://github.com/facebook/lexical/pull/7448/commits/0ee5a21f87126a3c68561abc8bbe504f76c2ca32)

5. [Fixed node replacement example](https://github.com/facebook/lexical/pull/7448/commits/650ac16d97d6db6b5c9007d66b2f21c3aeba3259):
   - Removed incorrect `replace()` method implementation
   - Added proper node replacement configuration using `editor.nodes`
   - Added reference to the Node Replacement guide

Please let me know if anything needs further adjustment!

etrepum commented on packages/lexical-website/docs/concepts/key-management.md:
```diff
@@ -0,0 +1,228 @@
+  __prev: null | NodeKey,
+  __next: null | NodeKey,
+  __parent: null | NodeKey,
+  __first: null | NodeKey,
+  __last: null | NodeKey
```
```suggestion
  // __first, __last and __size are only for ElementNode to track its children
  __first: null | NodeKey,
  __last: null | NodeKey,
  __size: number,
```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,175 @@
+
+```typescript
+// ✅ Correct: Modifying a node
+function updateNodeData(node: MyCustomNode, newData: string) {
+  node.getWritable().setData(newData);
+}
````
```suggestion
function $updateNodeData(node: MyCustomNode, newData: string): MyCustomNode {
  return node.setData(newData);
}
```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
```diff
@@ -0,0 +1,175 @@
+}
+
+// ❌ Incorrect: Don't clone manually
+function updateNodeDataWrong(node: MyCustomNode, newData: string) {
+  const clone = MyCustomNode.clone(node);
+  clone.setData(newData);
+  return clone;
+}
```
```suggestion
function $updateNodeDataWrong(node: MyCustomNode, newData: string): MyCustomNode {
  const clone = MyCustomNode.clone(node);
  clone.setData(newData);
  return clone;
}
```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,175 @@
+
+```typescript
+// ❌ Never do this
+function duplicateNode(node: MyCustomNode) {
````
```suggestion
function $duplicateNode(node: MyCustomNode) {
```

etrepum commented on packages/lexical-website/docs/concepts/node-cloning.md:
````diff
@@ -0,0 +1,175 @@
+
+```typescript
+// ✅ Correct: Copying a node
+function duplicateNode(node: MyCustomNode) {
+  return $copyNode(node);
+}
+
+// ❌ Incorrect: Don't use clone
+function duplicateNodeWrong(node: MyCustomNode) {
+  return node.constructor.clone(node);
+}
````
```suggestion
function $duplicateNode(node: MyCustomNode): MyCustomNode {
  return $copyNode(node);
}

// ❌ Incorrect: Don't use clone
function $duplicateNodeWrong(node: MyCustomNode): MyCustomNode {
  return node.constructor.clone(node);
}
```

etrepum commented on packages/lexical-website/docs/concepts/key-management.md:
````diff
@@ -0,0 +1,228 @@
+
+4. **Node Replacement**
+   ```typescript
+   // ❌ Never implement replace() directly in nodes
+   class MyCustomNode extends ElementNode {
+     replace(replaceWith: LexicalNode) {
+       return new DifferentNodeType(data, this.__key); // Wrong approach
+     }
+   }
````
```suggestion
   // ❌ Never re-use the key when changing the node class
   const editorConfig = {
     nodes: [
       CustomNodeType,
       {
         replace: OriginalNodeType,
         with: (node: OriginalNodeType) => new CustomNodeType(node.__key),
         withKlass: CustomNodeType
       }
     ]
   };
```

etrepum reviewed (commented):
It looks like I don't have write access to apply the suggestions but after these small edits this is good to go
